internal/controller/bucket_controller.go

```diff
@@ -0,0 +1,91 @@
+/*
+Copyright 2023.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/pkg/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/giantswarm/object-storage-operator/api/v1alpha1"
+	managementcluster "github.com/giantswarm/object-storage-operator/internal/pkg/managementcluster"
+	"github.com/giantswarm/object-storage-operator/internal/pkg/reconcilers/bucket"
+	"github.com/giantswarm/object-storage-operator/internal/pkg/reconcilers/bucket/capa"
+)
+
+// BucketReconciler reconciles a Bucket object
+type BucketReconciler struct {
+	client.Client
+	*runtime.Scheme
+	managementcluster.ManagementCluster
+}
+
+//+kubebuilder:rbac:groups=objectstorage.giantswarm.io,resources=buckets,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=objectstorage.giantswarm.io,resources=buckets/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=objectstorage.giantswarm.io,resources=buckets/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.16.0/pkg/reconcile
+func (r *BucketReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+
+	logger.Info("Started reconciling Bucket", "namespace", req.Namespace, "name", req.Name)
+	defer logger.Info("Finished reconciling Bucket", "namespace", req.Namespace, "name", req.Name)
+
+	// Get the bucket that we are reconciling
+	reconciledBucket := &v1alpha1.Bucket{}
+	err := r.Client.Get(ctx, req.NamespacedName, reconciledBucket)
+	if apierrors.IsNotFound(err) {
+		logger.Info("Bucket no longer exists")
+		return ctrl.Result{}, nil
+	} else if err != nil {
+		return ctrl.Result{}, errors.WithStack(err)
+	}
+
+	// Create the correct reconciler based on the provider
```

Comment:
what will be the differences between providers? I think it would be better to only add the complexity of several providers when that's needed, and I'd keep it simple now removing the provider variables and the switch.

Reply:
We will need it for azure blob storage for capz so I preferred to implément it now

Reply:
I don't see any benefits from doing it now that won't be used.

But anyway, with this design we will be duplicating logic between providers, like handling finalizers. Couldn't we use a "ObjectStorageService" that exposes methods like `Create()` or `Exists()` or `Delete()`, and use a different service implementation depending on the provider instead? That way the code is the same for all reconciled buckets, independently of the provider. We just create a different "ObjectStorageService" in the bucket reconciler depending on the provider: one for S3 while on AWS, one for blob storage while on Azure, etc.
Maybe there is a reason that prevents us from doing this that I'm not seeing as I haven't given a lot of thought to this problem.

Reply:
I think you're right. I think I wanted to move the creation of the s3 client outside the service but it does make sense, I'll rewrite that

Reply:
I changed the internal code by a lot to only have a service. I added a ConfigureBucket method to be able to differenciate Creation from Upgrades but that definitely better, thank you :)